flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/table/PrometheusDynamicSink.java

```diff
@@ -0,0 +1,193 @@
+import org.apache.flink.connector.prometheus.sink.PrometheusSinkConfiguration;
+import org.apache.flink.connector.prometheus.sink.prometheus.Types;
+import org.apache.flink.table.connector.ChangelogMode;
+import org.apache.flink.table.connector.sink.DynamicTableSink;
+import org.apache.flink.table.connector.sink.SinkV2Provider;
+import org.apache.flink.table.connector.sink.abilities.SupportsPartitioning;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.util.Preconditions;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Optional;
+import java.util.ServiceConfigurationError;
+import java.util.ServiceLoader;
+
+public class PrometheusDynamicSink extends AsyncDynamicTableSink<Types.TimeSeries>
+        implements SupportsPartitioning {
+    private static final Logger LOG = LoggerFactory.getLogger(PrometheusDynamicSink.class);
+
+    /** Consumed data type of the table. */
+    private final DataType physicalDataType;
+
+    private final PrometheusConfig prometheusConfig;
+
+    protected PrometheusDynamicSink(
+            @Nullable Integer maxBatchSize,
+            @Nullable Integer maxInFlightRequests,
+            @Nullable Integer maxBufferedRequests,
+            @Nullable Long maxBufferSizeInBytes,
+            @Nullable Long maxTimeInBufferMS,
+            @Nullable DataType physicalDataType,
+            PrometheusConfig prometheusConfig) {
+
+        super(
+                maxBatchSize, // maxBatchSizeInSamples
+                maxInFlightRequests,
+                maxBufferedRequests,
+                maxBufferSizeInBytes, // maxRecordSizeInSamples
+                maxTimeInBufferMS);
+
+        this.physicalDataType =
+                Preconditions.checkNotNull(physicalDataType, "Consumed data type must not be null");
+        this.prometheusConfig = prometheusConfig;
+    }
+
+    @Override
+    public ChangelogMode getChangelogMode(ChangelogMode changelogMode) {
+        return ChangelogMode.upsert();
+    }
+
+    @Override
+    public SinkRuntimeProvider getSinkRuntimeProvider(Context context) {
+
+        PrometheusSinkBuilder<RowData> builder =
+                new PrometheusSinkBuilder<RowData>()
+                        .setRetryConfiguration(
+                                PrometheusSinkConfiguration.RetryConfiguration
+                                        .DEFAULT_RETRY_CONFIGURATION)
+                        .setElementConverter(
+                                new RowDataElementConverter(physicalDataType, prometheusConfig))
+                        .setPrometheusRemoteWriteUrl(prometheusConfig.getRemoteWriteEndpointUrl());
+
+        Optional.ofNullable(prometheusConfig.getRequestSignerIdentifier())
+                .ifPresent(region -> builder.setRequestSigner(getRequestSigner(prometheusConfig)));
```

Comment:
I'm uncertain why `region` is being created within the context of a request signer identifier only not to be used in the subsequent call. Would this be simpler to leave what's required from the `prometheusConfig` to the `getRequestSigner` method which will return null if there isn't one and only call `builder.setRequestSigner` if it isn't?

Reply:
I agree, on top of your point, moving it to getRequestSigner is better here especially since we want to separate any Amazon config away from flink-connector-prometheus package. Updated and added unit test for this